starport/services/network/join_test.go

```diff
@@ -0,0 +1,370 @@
+package network
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/tendermint/starport/starport/pkg/cosmoserror"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	launchtypes "github.com/tendermint/spn/x/launch/types"
+	"github.com/tendermint/starport/starport/services/network/mocks"
+	"github.com/tendermint/starport/starport/services/network/testutil"
+)
+
+const (
+	TestDenom        = "stake"
+	TestAmountString = "95000000"
+	TestAmountInt    = int64(95000000)
+
+	TestAccountRequestID          = uint64(1)
+	TestGenesisValidatorRequestID = uint64(2)
+)
+
+func stubNetworkForJoin() Network {
+	launchQueryMock := new(mocks.LaunchClient)
+	launchQueryMock.On("GenesisValidator", mock.Anything, &launchtypes.QueryGetGenesisValidatorRequest{
+		Address:  testutil.Address,
+		LaunchID: testutil.TestLaunchID,
+	}).Return(nil, cosmoserror.ErrNotFound)
+	launchQueryMock.On("VestingAccount", mock.Anything, &launchtypes.QueryGetVestingAccountRequest{
+		Address:  testutil.Address,
+		LaunchID: testutil.TestLaunchID,
+	}).Return(nil, cosmoserror.ErrNotFound)
+	launchQueryMock.On("GenesisAccount", mock.Anything, &launchtypes.QueryGetGenesisAccountRequest{
+		Address:  testutil.Address,
+		LaunchID: testutil.TestLaunchID,
+	}).Return(nil, cosmoserror.ErrNotFound)
+	networkClientMock := new(mocks.CosmosClient)
+	networkClientMock.On(
+		"BroadcastTx",
+		testutil.AccountName,
+		mock.AnythingOfType("*types.MsgRequestAddValidator"),
+	).Return(testutil.NewResponse(&launchtypes.MsgRequestAddValidatorResponse{
+		RequestID:    TestGenesisValidatorRequestID,
+		AutoApproved: false,
+	}), nil)
+	networkClientMock.On(
+		"BroadcastTx",
+		testutil.AccountName,
+		mock.AnythingOfType("*types.MsgRequestAddAccount"),
+	).Return(testutil.NewResponse(&launchtypes.MsgRequestAddAccountResponse{
+		RequestID:    TestAccountRequestID,
+		AutoApproved: false,
+	}), nil)
+	return Network{
+		cosmos:      networkClientMock,
+		account:     testutil.GetTestAccount(),
+		launchQuery: launchQueryMock,
+	}
+}
+
+func TestJoin(t *testing.T) {
+	tmp := t.TempDir()
+	gentx := testutil.NewGentx(testutil.Address, TestDenom, TestAmountString, "")
+	gentxPath, err := gentx.SaveTo(tmp)
+	assert.Nil(t, err)
+	genesis := testutil.NewGenesis(testutil.TestChainChainID)
+	genesisPath, err := genesis.SaveTo(tmp)
+	assert.Nil(t, err)
+
+	network := stubNetworkForJoin()
+	chain := testutil.NewChainMock(testutil.WithGenesisPath(genesisPath), testutil.WithDefaultGentxPath(gentxPath))
+
+	joinErr := network.Join(context.Background(), chain, testutil.TestLaunchID, testutil.TestPublicAddress, "")
+	require.Nil(t, joinErr)
+	chain.AssertNumberOfCalls(t, "NodeID", 1)
```

Comment:
It's not enough to check number all of calls only, we should check the **arguments** given to these funcs and also what's being **returned**.